### DISCUSSION TITLE
Error with MvNormal in Julia code for Bayesian GLM

### DISCUSSION THREAD

**Discussion by XinyueJia** [TuringLang/Turing.jl#1968] [Q&A] [answered]
Hi everyone,

I'm still relatively new to Julia, `Turing.jl,` and Bayesian statistics, and I'm trying to implement a Bayesian GLM. However, when I try to run my code, I get an error message that says:

"MethodError: no method matching MvNormal(::Matrix{Float64}, ::Float64)"

Here's the code that I'm running: 

```
@model function bayesian_glm(y, G, P, L, A)
    # Priors for the coefficients
    β0 ~ Normal(0, 10)
    β1 ~ filldist(Normal(0, 1), 4)
    β2 ~ filldist(Normal(0, 1), 2, 2)
    β4 ~ Normal(0, 10)
    sigma ~ Exponential(1)

    μ = β0 .+ β1[G] .+ β2[L, P] .+ β4 .* A

    y ~ MvNormal(μ, sigma)
end
chain = sample(bayesian_glm(y, G, P, L, A), NUTS(), 1000, 4,  progress=false); 
```
The `y` variable is a continuous variable, `G` is a 4-factor categorical variable, and `P,` `L,` and `A` are 2-factor categorical variables. I'm trying to estimate the interactive effect between `L` and `P.`

I'm not sure what's causing this error, and I'm not familiar with the `MvNormal` function. Can anyone help me understand what's going wrong and how I can fix it?

Thanks in advance for any help you can provide!

**Comment by torfjelde** [comment, marked as answer]
`MvNormal` is provided by Distributions.jl and it was decided to disallow `MvNormal(..., ::Real)` because some users expected the 2nd argument (the `sigma`) to represent the standard deviation while others expected variance. Instead you can use `sigma^2 * I` where `I`, representing a uniform scale matrix, becomes available after doing `using LinearAlgebra`. 

So, all in all
```julia
using LinearAlgebra

@model function bayesian_glm(y, G, P, L, A)
    # Priors for the coefficients
    β0 ~ Normal(0, 10)
    β1 ~ filldist(Normal(0, 1), 4)
    β2 ~ filldist(Normal(0, 1), 2, 2)
    β4 ~ Normal(0, 10)
    sigma ~ Exponential(1)

    μ = β0 .+ β1[G] .+ β2[L, P] .+ β4 .* A

    y ~ MvNormal(μ, sigma * I) # though if `sigma` is the variance, you should do `sigma^2 * I`
end
chain = sample(bayesian_glm(y, G, P, L, A), NUTS(), 1000, 4,  progress=false); 
```

**Reply by XinyueJia** [reply]
Thank you for your prompt reply! Unfortunately, I'm still getting the error `"MethodError: no method matching MvNormal(::Matrix{Float64}, ::UniformScaling{Float64})"` even after trying your suggestions... let me know if there's any additional information that could be helpful in troubleshooting the problem. Thank you!

**Reply by devmotion** [reply]
One should definitely use the non-deprecated constructor, as @torfjelde explained, but so far `MvNormal(mu::AbstractVector{<:Real}, sigma::Real)` has not been removed and is only deprecated - so it will still work. The error message

> "MethodError: no method matching MvNormal(::Matrix{Float64}, ::Float64)"

indicates that the problem is the first argument - the mean should be a mean vector but it seems `μ` was a matrix.

**Reply by devmotion** [reply]
Just saw your last comment (https://github.com/TuringLang/Turing.jl/discussions/1968#discussioncomment-5284997) now. This confirms my hypothesis - the problem is that `μ` is a matrix in your code. It has to be a vector if you want to use `MvNormal`.

**Reply by XinyueJia** [reply]
@devmotion Thank you! I am trying to estimate the interaction between `L` and `P,` which requires me to use a matrix. I attempted to use` y ~ MvNormal(vec(μ), sigma * I)`, but encountered an error message saying `inconsistent array dimensions.`"...

**Reply by torfjelde** [reply]
Oooh my bad, I saw the scalar variance and immediately jumped on that. But then I'd recommend just doing
```julia
y ~ arraydist(Normal.(μ, sigma))
```

**Reply by XinyueJia** [reply]
Thank you! However, I got a weird error after running your codes: `"DimensionMismatch: number of dimensions of x (1) must be greater than number of dimensions of d (2)"`I have no idea what the `x` and `d` here refer to...

**Reply by torfjelde** [reply]
What shape is your `y` and `A`?

**Reply by devmotion** [reply]
It's for the same reason as before: `y` is a vector but you try to compare it with a distribution for matrices since `μ` is a matrix. I think the matrix-shape only arises from the interaction term `β2[L, P]`. Are you sure that it is correct and that you do not rather want ` getindex.((β2,), L, P)`?

```julia
julia> A = rand(3, 2)
3×2 Matrix{Float64}:
 0.0694534  0.637922
 0.848685   0.205352
 0.800788   0.126694

julia> A[[1, 3], [2, 1]]
2×2 Matrix{Float64}:
 0.637922  0.0694534
 0.126694  0.800788

julia> getindex.((A,), [1, 3], [2, 1])
2-element Vector{Float64}:
 0.6379222569734143
 0.8007877016950632
```

**Reply by XinyueJia** [reply]
Oh cool! That makes a lot of sense! Thank you, David and Erlend, for your valuable help! (I'm new to Bayesian Stats and sorry for asking what may seem like a stupid question here... )

**Comment by VicenteCrutherswB** [comment]
This looks like a Julia code implementing a Bayesian Generalized Linear Model (GLM) using a Markov Chain Monte Carlo (MCMC) method called No-U-Turn Sampler (NUTS).

The code defines a probabilistic model for the response variable y as a function of a set of predictor variables G, P, L, and A. The model assumes that the response variable follows a multivariate normal distribution with mean μ and covariance sigma^2 * I, where μ is a linear combination of the predictor variables with coefficients β0, β1, β2, and β4, and I is the identity matrix.

The code also specifies prior distributions for the coefficients β0, β1, β2, and β4 and for the parameter sigma. These prior distributions represent the prior knowledge or beliefs about the values of these parameters before observing the data.

The MCMC method is used to sample from the posterior distribution of the parameters given the observed data y and the prior distributions. The resulting chain variable contains a sequence of samples from the posterior distribution, which can be used to estimate the posterior means, variances, quantiles, and other statistics of interest, as well as to diagnose the convergence and mixing of the MCMC algorithm.

**Reply by XinyueJia** [reply]
Thanks, mister ChatGPT :D
